### PR TITLE
[#174210092] Create Coll-8 gem with fetch auth token

### PR DESCRIPTION
**Pull request by Eddie901**
https://www.pivotaltracker.com/story/show/174210092

## Why
We separate the carrier specific logic from our application to a Gem.

## Details
We need to create a Gem for Coll-8 and implement the following operations

- Get token
- Generic operation class for real operation to inherit from

To access their Ship API we need to require access_token. The Ship API uses OAuth 2.0 to authenticate and authorise requests. Details for getting a token are provided in the API integration link.
We need to get the test credentials from coll-8 to perform testing

Coll-8 provides different services for creating a shipment. Reference is given in following [link]
(https://developerportal.drop2shop.ie/API-Integration/service-types.pdf)
We need to use drop2me (Shipping from Retailer to Consumer home address) shipment service for creating a shipment.

As Developer
I want to a basic Coll-8 gem which fetches the auth token
So that I can add the operations in the next ticket

Notes
We should use UAT (User Acceptance Test) to fetch the authentication token
Please see link for API documentation.

!!!!!!!! Make sure to not include our test credentials as part of testing cassettes !!!!!!!!

### CODE REVIEW

File: lib/coll_8_api/auth_token.rb [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+      http_client = Faraday.new
+
+      response = http_client.post(TOKEN_ENDPOINT) do |req|
+        req.headers['Content-Type'] = 'application/json'
+        req.body = payload.to_json
+      end
+
+      body = JSON.parse(response.body)
+
+      raise StandardError.new(payload: payload, body: body, status: response.status) unless response.success?
```
Comment (gdackys) [inline lib/coll_8_api/auth_token.rb]: Should this be `ResponseError` instead?

File: lib/coll_8_api/auth_token.rb [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+      body
+    end
+
+    def payload
+      {
+        "client_id": config.client_id,
+        "client_secret": config.client_secret,
+        "audience": AUDIENCE,
+        "grant_type": GRANT_TYPE
```
Comment (gdackys) [inline lib/coll_8_api/auth_token.rb:35]: No need for double quotes here.

File: lib/coll_8_api/errors/response_error.rb [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+        super(build_message)
+      end
+
+      private
+
+      def build_message
+        return status unless body.is_a?(Hash)
+
+        body.dig(:reasons, 0, :msg) || status
```
Comment (gdackys) [inline lib/coll_8_api/errors/response_error.rb]: Are we certain that Coll8 returns this erroneous response structure?

File: lib/coll_8_api/auth_token.rb (hunk 2 above)
Comment (Eddie901) [reply lib/coll_8_api/auth_token.rb:35]: rubocop doesn't complain about those ones. I think double quotes is normal for hash keys, no?

File: lib/coll_8_api/errors/response_error.rb (hunk 3 above)
Comment (Eddie901) [reply lib/coll_8_api/errors/response_error.rb]: Will return simple status until with have concrete examples of the error messaging.

File: lib/coll_8_api/auth_token.rb (hunk 2 above)
Comment (gdackys) [reply lib/coll_8_api/auth_token.rb:35]: It depends on what the hash key is supposed to be: if a symbol then quotes are unnecessary unless it contains non-alphanumeric characters or interpolation is needed.

**Review by gdackys — APPROVED** [review]

**Pull request merged**